Add tagging rule to history level rule builder, copy HDAs when building collection

mvdbeek opened the pull request:
Tagging is useful also when you already have a history with datasets that contain for example group information in the name. To avoid adding tags to the source HDA, which might be confusing we now copy HDAs when building collections from HDAs in the UI, but default to not doing this in the API for backwards compatibility.

mvdbeek committed:
Add tag and group tag rule to history level rule builder

jmchilton commented:
Does this work? I'm haven't tried this - but I don't understand how it could work without updating the dataset collection creation API to attach tags as it creates the collections. Also I don't think that API endpoint clones the HDAs so you are going to attach the tags to existing HDAs - that might be confusing?

It is a good idea - I think we want to duplicate the HDAs though or issue a warning about this or something?

mvdbeek commented:
Totally doesn't work, should have put this in wip

mvdbeek committed:
Reuse HDA copy method in HDAManager

Should benefit from smarter flush strategy in HDA.copy()

mvdbeek committed:
Always copy HDAs when creating a new collection from exisitng datasets, and allow setting tags

mvdbeek committed:
Test creation of collection from existing dataset, with a new tag

mvdbeek committed:
Simplify hda manager usage

mvdbeek committed:
Include tags in element_identifier when building collection from datasets

mvdbeek committed:
Restore hiding of dataset instances

mvdbeek committed:
Adapt unit tests to copying HDAs when creating collections

mvdbeek committed:
Adjust API tests to copying HDAs when creating collection

mvdbeek committed:
Treat LDDA as HDA when creating collection

This avoids copying the LDDA to HDA and then copying the HDA again.

mvdbeek committed:
Test for proper 403 response

Coincidentally fixed by e013f2fbd1596b103ef195afcba8cbd787c49b33.

mvdbeek committed:
Default to not copying source HDAs when creating new collections via API

mvdbeek committed:
Revert "Adapt unit tests to copying HDAs when creating collections"

This reverts commit 408a003f171be73e4b347579ee6f662a18fc1990.

mvdbeek committed:
Revert "Adjust API tests to copying HDAs when creating collection"

This reverts commit f120390a63700ff0ec6c09185aad26b91ad8f842.

mvdbeek committed:
Send copy_elements=True in new create collection test

mvdbeek committed:
Default to copying HDAs when  creating collections via the UI

mvdbeek committed:
Add unit test for copy_element option

mvdbeek commented:
Hmm, somethings not quite right here. For this rule:
```
{
  "rules": [
    {
      "type": "add_column_metadata",
      "value": "name"
    },
    {
      "type": "add_column_regex",
      "target_column": 0,
      "expression": "(.*)-..",
      "replacement": "\u0000genotype:\\1"
    },
    {
      "type": "add_column_regex",
      "target_column": 0,
      "expression": ".*-(.).",
      "replacement": "\u0000age:\\1"
    },
    {
      "type": "add_column_regex",
      "target_column": 0,
      "expression": ".*-.(.)",
      "replacement": "\u0000rep:\\1"
    }
  ],
  "mapping": [
    {
      "type": "list_identifiers",
      "columns": [
        0
      ],
      "editing": false
    }
  ]
}
```
I'm seeing:
```
Traceback (most recent call last):
  File "lib/galaxy/web/framework/decorators.py", line 283, in decorator
    rval = func(self, trans, *args, **kwargs)
  File "lib/galaxy/webapps/galaxy/api/history_contents.py", line 396, in create
    return self.__create_dataset_collection(trans, history, payload, **kwd)
  File "lib/galaxy/webapps/galaxy/api/history_contents.py", line 525, in __create_dataset_collection
    **create_params
  File "lib/galaxy/managers/collections.py", line 108, in create
    copy_elements=copy_elements,
  File "lib/galaxy/managers/collections.py", line 181, in create_dataset_collection
    copy_elements=copy_elements)
  File "lib/galaxy/managers/collections.py", line 217, in _element_identifiers_to_elements
    copy_elements=copy_elements)
  File "lib/galaxy/managers/collections.py", line 389, in __load_elements
    copy_elements=copy_elements)
  File "lib/galaxy/managers/collections.py", line 434, in __load_element
    self.tag_manager.apply_item_tags(user=trans.user, item=element, tags_str=tag_str)
  File "lib/galaxy/managers/tags.py", line 182, in apply_item_tags
    self.apply_item_tag(user, item, name, value)
  File "lib/galaxy/managers/tags.py", line 154, in apply_item_tag
    tag = self._get_or_create_tag(lc_name)
  File "lib/galaxy/managers/tags.py", line 233, in _get_or_create_tag
    tag = self.get_tag_by_name(scrubbed_tag_str)
  File "lib/galaxy/managers/tags.py", line 205, in get_tag_by_name
    return self.sa_session.query(galaxy.model.Tag).filter_by(name=tag_name.lower()).first()
  File "/bioinfo/guests/mvandenb/galaxy/.venv/local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2789, in first
    ret = list(self[0:1])
  File "/bioinfo/guests/mvandenb/galaxy/.venv/local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2581, in __getitem__
    return list(res)
  File "/bioinfo/guests/mvandenb/galaxy/.venv/local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2889, in __iter__
    return self._execute_and_instances(context)
  File "/bioinfo/guests/mvandenb/galaxy/.venv/local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2912, in _execute_and_instances
    result = conn.execute(querycontext.statement, self._params)
  File "/bioinfo/guests/mvandenb/galaxy/.venv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 948, in execute
    return meth(self, multiparams, params)
  File "/bioinfo/guests/mvandenb/galaxy/.venv/local/lib/python2.7/site-packages/sqlalchemy/sql/elements.py", line 269, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/bioinfo/guests/mvandenb/galaxy/.venv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1060, in _execute_clauseelement
    compiled_sql, distilled_params
  File "/bioinfo/guests/mvandenb/galaxy/.venv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1200, in _execute_context
    context)
  File "/bioinfo/guests/mvandenb/galaxy/.venv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1416, in _handle_dbapi_exception
    util.reraise(*exc_info)
  File "/bioinfo/guests/mvandenb/galaxy/.venv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1193, in _execute_context
    context)
  File "/bioinfo/guests/mvandenb/galaxy/.venv/local/lib/python2.7/site-packages/sqlalchemy/engine/default.py", line 507, in do_execute
    cursor.execute(statement, parameters)
ValueError: A string literal cannot contain NUL (0x00) characters.
```

jmchilton commented on lib/galaxy/managers/collections.py:
```diff
@@ -399,11 +416,24 @@ def __load_element(self, trans, element_identifier):
+            else:
+                element = hda
+            if hide_source_items:
+                hda.visible = False
```
This would need to be owned not just accessible right?

mvdbeek replied:
Oh yeah, good point!

mvdbeek committed:
Original HDA must be owned by user if hiding source elements

mvdbeek committed:
Strip unicode control characters from tags

Leaving them in can trip up postgresql:
```
Traceback (most recent call last):
  File "lib/galaxy/web/framework/decorators.py", line 283, in decorator
    rval = func(self, trans, *args, **kwargs)
  File "lib/galaxy/webapps/galaxy/api/history_contents.py", line 396, in create
    return self.__create_dataset_collection(trans, history, payload, **kwd)
  File "lib/galaxy/webapps/galaxy/api/history_contents.py", line 525, in __create_dataset_collection
    **create_params
  File "lib/galaxy/managers/collections.py", line 108, in create
    copy_elements=copy_elements,
  File "lib/galaxy/managers/collections.py", line 181, in create_dataset_collection
    copy_elements=copy_elements)
  File "lib/galaxy/managers/collections.py", line 217, in _element_identifiers_to_elements
    copy_elements=copy_elements)
  File "lib/galaxy/managers/collections.py", line 389, in __load_elements
    copy_elements=copy_elements)
  File "lib/galaxy/managers/collections.py", line 434, in __load_element
    self.tag_manager.apply_item_tags(user=trans.user, item=element, tags_str=tag_str)
  File "lib/galaxy/managers/tags.py", line 182, in apply_item_tags
    self.apply_item_tag(user, item, name, value)
  File "lib/galaxy/managers/tags.py", line 154, in apply_item_tag
    tag = self._get_or_create_tag(lc_name)
  File "lib/galaxy/managers/tags.py", line 233, in _get_or_create_tag
    tag = self.get_tag_by_name(scrubbed_tag_str)
  File "lib/galaxy/managers/tags.py", line 205, in get_tag_by_name
    return self.sa_session.query(galaxy.model.Tag).filter_by(name=tag_name.lower()).first()
  File "/bioinfo/guests/mvandenb/galaxy/.venv/local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2789, in first
    ret = list(self[0:1])
  File "/bioinfo/guests/mvandenb/galaxy/.venv/local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2581, in __getitem__
    return list(res)
  File "/bioinfo/guests/mvandenb/galaxy/.venv/local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2889, in __iter__
    return self._execute_and_instances(context)
  File "/bioinfo/guests/mvandenb/galaxy/.venv/local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2912, in _execute_and_instances
    result = conn.execute(querycontext.statement, self._params)
  File "/bioinfo/guests/mvandenb/galaxy/.venv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 948, in execute
    return meth(self, multiparams, params)
  File "/bioinfo/guests/mvandenb/galaxy/.venv/local/lib/python2.7/site-packages/sqlalchemy/sql/elements.py", line 269, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/bioinfo/guests/mvandenb/galaxy/.venv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1060, in _execute_clauseelement
    compiled_sql, distilled_params
  File "/bioinfo/guests/mvandenb/galaxy/.venv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1200, in _execute_context
    context)
  File "/bioinfo/guests/mvandenb/galaxy/.venv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1416, in _handle_dbapi_exception
    util.reraise(*exc_info)
  File "/bioinfo/guests/mvandenb/galaxy/.venv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1193, in _execute_context
    context)
  File "/bioinfo/guests/mvandenb/galaxy/.venv/local/lib/python2.7/site-packages/sqlalchemy/engine/default.py", line 507, in do_execute
    cursor.execute(statement, parameters)
ValueError: A string literal cannot contain NUL (0x00) characters.
```

mvdbeek commented:
OK, I figured out the problem with the "weird characters", these are unicode control characters. In 366701c I'm stripping these in the parsing the tag string. So that should close #6600 as well.

jmchilton commented on test/unit/managers/test_TagManager.py:
```diff
@@ -0,0 +1,107 @@
+# -*- coding: utf-8 -*-
```
Amazing!
